### PR TITLE
8334328: Reduce object allocation for FloatToDecimal and DoubleToDecimal

### PR DESCRIPTION
**Pull request by wenshao**
The current versions of FloatToDecimal and DoubleToDecimal allocate additional objects. Reducing these allocations can improve the performance of Float/Double.toString and AbstractStringBuilder's append(float/double).

This patch is just a code refactoring to reduce object allocation, but does not change the Float/Double to decimal algorithm.

The following code comments the allocated objects to be removed.

```java
class FloatToDecimal {
    public static String toString(float v) {
        // allocate object FloatToDecimal
        return new FloatToDecimal().toDecimalString(v);
    }

    public static Appendable appendTo(float v, Appendable app)
            throws IOException {
        // allocate object FloatToDecimal
        return new FloatToDecimal().appendDecimalTo(v, app);
    }

    private Appendable appendDecimalTo(float v, Appendable app)
            throws IOException {
        switch (toDecimal(v)) {
            case NON_SPECIAL:
                // allocate object char[]
                char[] chars = new char[index + 1];
                for (int i = 0; i < chars.length; ++i) {
                    chars[i] = (char) bytes[i];
                }
                if (app instanceof StringBuilder builder) {
                    return builder.append(chars);
                }
                if (app instanceof StringBuffer buffer) {
                    return buffer.append(chars);
                }
                for (char c : chars) {
                    app.append(c);
                }
                return app;
            // ...
        }
    }
}

class DoubleToDecimal {
    public static String toString(double v) {
        // allocate object DoubleToDecimal
        return new DoubleToDecimal(false).toDecimalString(v);
    }

    public static Appendable appendTo(double v, Appendable app)
            throws IOException {
        // allocate object DoubleToDecimal
        return new DoubleToDecimal(false).appendDecimalTo(v, app);
    }

    private Appendable appendDecimalTo(double v, Appendable app)
            throws IOException {
        switch (toDecimal(v, null)) {
            case NON_SPECIAL:
                // allocate object char[]
                char[] chars = new char[index + 1];
                for (int i = 0; i < chars.length; ++i) {
                    chars[i] = (char) bytes[i];
                }
                if (app instanceof StringBuilder builder) {
                    return builder.append(chars);
                }
                if (app instanceof StringBuffer buffer) {
                    return buffer.append(chars);
                }
                for (char c : chars) {
                    app.append(c);
                }
                return app;
            // ...
        }
    }
}
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334328](https://bugs.openjdk.org/browse/JDK-8334328): Reduce object allocation for FloatToDecimal and DoubleToDecimal (**Enhancement** - P4)

### Reviewers
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**)
 * [Raffaello Giulietti](https://openjdk.org/census#rgiulietti) (@rgiulietti - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19730/head:pull/19730` \
`$ git checkout pull/19730`

Update a local copy of the PR: \
`$ git checkout pull/19730` \
`$ git pull https://git.openjdk.org/jdk.git pull/19730/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19730`

View PR using the GUI difftool: \
`$ git pr show -t 19730`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19730.diff">https://git.openjdk.org/jdk/pull/19730.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19730#issuecomment-2169041557)

### CODE REVIEW

**Comment by wenshao** [comment]
The performance numbers under `MacBookPro M1 Pro` & `MacBookPro 2018 i9` are as follows:

## Apple MacBookPro 2018
* CPU Intel i9 (x64)
```
-Benchmark                               Mode  Cnt    Score   Error  Units
-StringBuilders.appendWithFloat8Latin1   avgt   15  381.206 ? 8.681  ns/op
-StringBuilders.appendWithFloat8Utf16    avgt   15  366.549 ? 1.775  ns/op
-StringBuilders.appendWithDouble8Latin1  avgt   15  570.085 ? 8.040  ns/op
-StringBuilders.appendWithDouble8Utf16   avgt   15  501.409 ? 5.750  ns/op

+Benchmark                               Mode  Cnt    Score    Error  Units (Webrevs 00 4c810154)
+StringBuilders.appendWithFloat8Latin1   avgt   15  226.236 ?  3.350  ns/op +68.49%
+StringBuilders.appendWithFloat8Utf16    avgt   15  299.072 ? 19.895  ns/op +22.56%
+StringBuilders.appendWithDouble8Latin1  avgt   15  328.887 ?  5.865  ns/op +73.33%
+StringBuilders.appendWithDouble8Utf16   avgt   15  390.618 ? 30.142  ns/op +28.36%
```

## MacBook M1 Pro
* CPU Apple M1 Pro (aarch64)
```
-Benchmark                               Mode  Cnt    Score    Error  Units base
-StringBuilders.appendWithFloat8Latin1   avgt   15  317.144 ? 11.325  ns/op
-StringBuilders.appendWithFloat8Utf16    avgt   15  316.980 ? 17.955  ns/op
-StringBuilders.appendWithDouble8Latin1  avgt   15  440.853 ? 13.067  ns/op
-StringBuilders.appendWithDouble8Utf16   avgt   15  418.896 ?  4.610  ns/op


+Benchmark                               Mode  Cnt    Score    Error  Units (Webrevs 00 4c810154)
+StringBuilders.appendWithFloat8Latin1   avgt   15  168.231 ?  4.749  ns/op +88.51%
+StringBuilders.appendWithFloat8Utf16    avgt   15  213.981 ?  3.274  ns/op +48.13%
+StringBuilders.appendWithDouble8Latin1  avgt   15  241.536 ?  0.993  ns/op +82.52%
+StringBuilders.appendWithDouble8Utf16   avgt   15  284.863 ? 10.381  ns/op +47.05%

```

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:44]: We should copy these comments to `ToDecimal`; otherwise, it's not obvious what references are being made in comments there.

File: src/java.base/share/classes/java/lang/AbstractStringBuilder.java
Comment (wenshao) [reply src/java.base/share/classes/java/lang/AbstractStringBuilder.java:888]: The maximum length of float to decimal is 15, and the maximum length of double to decimal is 24. The original exception handling was because the Appendable#append(char) method threw an IOException, which is no longer necessary.

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/math/ToDecimal.java]: I have used JavaLangAccess instead of Usnafe

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (cl4es) [inline src/java.base/share/classes/jdk/internal/math/ToDecimal.java:171]: This should be named something like `asciiBytesToString` now. Perhaps `ToDecimal.LATIN1` can be renamed `ASCII`, too.

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (cl4es) [inline src/java.base/share/classes/jdk/internal/math/ToDecimal.java:169]: Enhances performance over what, `new String(str, 0, index, StandardCharsets.US_ASCII)`? Please specify.

File: test/micro/org/openjdk/bench/java/lang/StringBuilders.java
Comment (cl4es) [inline test/micro/org/openjdk/bench/java/lang/StringBuilders.java:244]: It would be interesting with a mixed micro to explore effects of increased number of implementors (looping over an array with `sbLatin1` and `sbUtf16`, and appending both floats and doubles in the same loop) 

`buf.delete(0, buf.length())` could be replaced with `buf.setLength(0)` in these. Might reduce a bit of overhead and focus micros .

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (cl4es) [inline src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:185]:
```suggestion
     *  Returns size in the lower byte, type in the high byte, where type is
```

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (cl4es) [inline src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:190]:
```suggestion
     *     NAN             iff v is NaN
     *     otherwise NON_SPECIAL
```

**Review by cl4es — CHANGES_REQUESTED** [review]
Just suggesting some improvements

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/math/ToDecimal.java:169]: The previous version of the code is like this, which will affect Float.toString and Double.toString. If changed, it is necessary to add the Benchmark of Float.toString and Double.toString, so I don’t want to change this in this PR.

Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/math/ToDecimal.java]: I followed your suggestion and changed it, but Coding uses ISO_8859_1. The performance of ISO_8859_1 is faster than that of the previously abandoned method.

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/math/ToDecimal.java]: In JDK internal string processing, LATIN1 and UTF16 appear in pairs, so I think LATIN1 is better.

**Comment by wenshao** [comment]
All suggestions have been fixed, can this PR be integrated? @cl4es  @liach

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/math/ToDecimal.java:154]: Can we use `s.getBytes(0, length, str, index)` and add `@SuppressWarnings("deprecation")`

**Review by liach — COMMENTED** [review]
I would request a review from @rgiulietti, author of the original code.

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (rgiulietti) [inline src/java.base/share/classes/jdk/internal/math/ToDecimal.java:34]: I think this should better be

```suggestion
abstract sealed class ToDecimal permits DoubleToDecimal, FloatToDecimal {
```

since we don't want/need to instantiate this class.

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (rgiulietti) [inline src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:175]:
```suggestion
        assert 0 <= index && index <= length(str) - MAX_CHARS : "Trusted caller missed bounds check";
```
This avoids a potential overflow.
But if you want to enforce the check, you should probably avoid using `assert` since assertions might be disabled.

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (rgiulietti) [inline src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:43]:
```suggestion
     * Use LATIN1 encoding to process the in-out byte[] str
```

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (rgiulietti) [inline src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:49]:
```suggestion
     * Use UTF16 encoding to process the in-out byte[] str
```

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (rgiulietti) [inline src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:236]: I suggest restoring the original logic like so:
```
                /* subnormal value */
                return (t < C_TINY
                        ? toDecimal(str, index, Q_MIN, 10 * t, -1, fd)
                        : toDecimal(str, index, Q_MIN, t, 0, fd)) - start;
```

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (rgiulietti) [inline src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:452]: As discussed in `ToDecimal`, all `put...` methods should return the updated `index`. Then this code can be rewritten as
```suggestion
            index = put8Digits(str, index, l);
```
so that the caller doesn't need to know how many characters were added.

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (rgiulietti) [inline src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:460]: By the same logic as suggested above for `put8Digits`
```suggestion
            index = putChar(str, index, '-');
```

and other places as well.

File: src/java.base/share/classes/jdk/internal/math/FloatToDecimal.java
Comment (rgiulietti) [inline src/java.base/share/classes/jdk/internal/math/FloatToDecimal.java:41]: The review of this class is the same as for `DoubleToDecimal`, so I won't repeat it here.

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (rgiulietti) [inline src/java.base/share/classes/jdk/internal/math/ToDecimal.java:45]: These constants could be like the original ones (0, 1, 2, ...).
This would make the switch over them a `tableswitch` rather than a `lookupswitch` in bytecode. I'm not sure whether this would improve machine code, however.
Of course, logic would need to be changed in some places (`<< 8`, `>>> 8`, etc.)

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (rgiulietti) [inline src/java.base/share/classes/jdk/internal/math/ToDecimal.java:63]: I think it would be better for all the `put...` methods here to return the updated `index`, hence replacing the return type from `void` to `int`.

This would make usages easier, as there's no need to remember at the call sites by how much the index must be incremented after the usage.

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (rgiulietti) [inline src/java.base/share/classes/jdk/internal/math/ToDecimal.java:71]: This method is used to put at least one digit. This means that the caller has to remember to add a '0' explicitly, which is questionable.
I suggest removing this method and invoke the correct combination of `putChar` and `putDigit` for more readable code.

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (rgiulietti) [inline src/java.base/share/classes/jdk/internal/math/ToDecimal.java:165]:
```suggestion
    final int length(byte[] str) {
```

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:236]: I like the new implementation, the code is cleaner, is your suggestion to revert to the original version due to smaller changes?

**Comment by rgiulietti** [comment]
@wenshao Looks good, thanks for the improvements.

Let me know when you are finished with your changes.
Once approved, you should refrain from adding other commits without further approvals.

**Comment by wenshao** [comment]
@rgiulietti 
Thanks for your review. I have completed all the work I wanted to do on this PR. I will not make any changes without further suggestions.

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (rgiulietti) [reply src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:236]: In principle, you should not arbitrarily change code that is correct, and only limit your changes to reach the goal of the PR.
My suggestion is the minimal change required, yours is more than strictly needed.

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (cl4es) [inline src/java.base/share/classes/jdk/internal/math/ToDecimal.java:47]: I think this somewhat unnecessarily copies names and internal implementation details from `String` (where the `coder` value is used both as a pseudo-boolean and as a shift operand, which is not applicable here). In this context we could use something like `private final boolean latin1;` (all uses of `coder` are just `if (coder == LATIN1)` so it would be simplified to `if (latin1)`)

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:236]: Got it, should I change it back?

Comment (rgiulietti) [reply src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:236]: If there's nothing else, no, there's no need to switch back.

Sorry if I sound pedantic on some principles that I feel are important for the good health of OpenJDK ;-)

File: src/java.base/share/classes/jdk/internal/math/ToDecimal.java
Comment (cl4es) [reply src/java.base/share/classes/jdk/internal/math/ToDecimal.java:47]: Yeah, I'm just musing about the perils of leaking/replicating those implementation details, but as you say this is internal code only used by `String` and `StringBuilder` to it's loosely part of that complex. Not requesting any changes here for now..

Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/math/ToDecimal.java:47]: I looked at it again and I think @cl4es 's suggestion is correct. Can I submit the change? @rgiulietti

Comment (rgiulietti) [reply src/java.base/share/classes/jdk/internal/math/ToDecimal.java:47]: Of course, nothing has been approved as of now.

Since you are preparing a commit anyway, may I ask you to revert back the changes [here](https://github.com/openjdk/jdk/pull/19730#discussion_r1652978056)? Thanks.

Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/math/ToDecimal.java]: I have completed all the suggested changes. Please continue to review.

**Review by rgiulietti — APPROVED** [review]
LGTM

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:194]: This `toDecimal` returns packed int with info but the other returns the new index; this is extremely confusing for future readers, especially that they have very similar signatures (difference in signature is in the middle). I recommend making this return a `long` so when it's downcasted to an int, it's just the size (like String concat's coderIndex)

**Review by liach — CHANGES_REQUESTED** [review]

**Review by cl4es — APPROVED** [review]
LGTM2

File: src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java
Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java:194]: The maximum length of float to decimal is 15, and the maximum length of double to decimal is 24, so one byte is enough. If we use the long return type, we have to add `cast int to long` and `cast long to int` to the code, which is redundant. It will also confuse people, thinking that float to decimal and double to decimal will have a large length, just like the size of string concat, which requires a complete int.

**Comment by liach** [comment]
Tier 1-3 tests look good. Thanks Wen Shaojin!

/sponsor

**Pull request closed without merging**
